File: StrikeTrack-Expo/lib/batteryBeakParser.ts
import type { BatteryStatus } from './constants';

export type ParsedBatteryReading = {
  status?: BatteryStatus;
  chargePercent?: number;
  voltageNoLoad?: number;
  voltageLoad1?: number;
  voltageLoad2?: number;
  currentLoad2?: number;
  internalResistance?: number;
};

const STATUSES: BatteryStatus[] = ['Good', 'Fair', 'Bad', 'Charge Battery'];

function parseStatus(text: string): BatteryStatus | undefined {
  const match = text.match(/Status:\s*(Good|Fair|Bad|Charge Battery)/i);
  if (match) {
    const s = STATUSES.find((x) => x.toLowerCase() === match[1].toLowerCase());
    if (s) return s;
  }
  for (const status of STATUSES) {
    if (text.includes(status)) return status;
  }
  return undefined;
}

function parseCharge(text: string): number | undefined {
  const match = text.match(/Charge:\s*(\d{1,3})%?/i);
  if (match) {
    const val = parseFloat(match[1]);
    return Math.min(130, Math.max(0, val));
  }
  const pctMatch = text.match(/\d{1,3}%/);
  if (pctMatch) {
    const val = parseFloat(pctMatch[0].replace('%', ''));
    return Math.min(130, Math.max(0, val));
  }
  return undefined;
}

function parseVoltageLine(
  text: string,
  prefix: string
): [number | undefined, number | undefined] {
  const pattern = new RegExp(
    `${prefix}:\\s*([0-9.]+)\\s*([0-9]+)\\s*Amps?`,
    'i'
  );
  const match = text.match(pattern);
  if (match) {
    const voltage = parseFloat(match[1]);
    const amps = match[2] ? parseFloat(match[2]) : undefined;
    return [voltage, amps];
  }
  return [undefined, undefined];
}

function parseRint(text: string): number | undefined {
  const match = text.match(/Rint:\s*([\d.]+)\s*Ohms?/i);
  if (match) {
    return parseFloat(match[1]);
  }
  const ohmsMatch = text.match(/[\d.]+\s*Ohms?/i);
  if (ohmsMatch) {
    const numStr = ohmsMatch[0].replace(/[^0-9.]/g, '');
    const val = parseFloat(numStr);
    if (!isNaN(val)) return val;
  }
  return undefined;
}

export function parseBatteryBeak(ocrText: string): ParsedBatteryReading {
  const text = ocrText
    .replace(/\n/g, ' ')
    .replace(/\s{2,}/g, ' ')
    .trim();

  const [v0] = parseVoltageLine(text, 'V0');
  const [v1] = parseVoltageLine(text, 'V1');
  const [, amps2] = parseVoltageLine(text, 'V2');

  return {
    status: parseStatus(text),
    chargePercent: parseCharge(text),
    voltageNoLoad: v0,
    voltageLoad1: v1,
    voltageLoad2: parseVoltageLine(text, 'V2')[0],
    currentLoad2: amps2,
    internalResistance: parseRint(text),
  };
}

File: StrikeTrack/Services/BatteryBeakParser.swift
import Foundation

struct ParsedBatteryReading {
    var status: BatteryStatus?
    var chargePercent: Double?
    var voltageNoLoad: Double?
    var voltageLoad1: Double?
    var voltageLoad2: Double?
    var currentLoad2: Double?
    var internalResistance: Double?
}

struct BatteryBeakParser {
    static func parse(_ ocrText: String) -> ParsedBatteryReading {
        var result = ParsedBatteryReading()
        let text = ocrText
            .replacingOccurrences(of: "\n", with: " ")
            .replacingOccurrences(of: "  ", with: " ")
            .trimmingCharacters(in: .whitespaces)

        result.status = parseStatus(from: text)
        result.chargePercent = parseCharge(from: text)
        let (v0, _) = parseVoltageLine(from: text, prefix: "V0")
        let (v1, _) = parseVoltageLine(from: text, prefix: "V1")
        let (v2, amps2) = parseVoltageLine(from: text, prefix: "V2")
        result.voltageNoLoad = v0
        result.voltageLoad1 = v1
        result.voltageLoad2 = v2
        result.currentLoad2 = amps2
        result.internalResistance = parseRint(from: text)

        return result
    }

    private static func parseStatus(from text: String) -> BatteryStatus? {
        if let match = text.range(of: #"Status:\s*(Good|Fair|Bad|Charge Battery)"#, options: .regularExpression) {
            let substr = String(text[match])
            if let statusStr = substr.split(separator: ":").last?.trimmingCharacters(in: .whitespaces),
               let status = BatteryStatus(rawValue: statusStr) {
                return status
            }
        }
        for status in BatteryStatus.allCases {
            if text.contains(status.rawValue) {
                return status
            }
        }
        return nil
    }

    private static func parseCharge(from text: String) -> Double? {
        if let match = text.range(of: #"Charge:\s*(\d{1,3})%?"#, options: .regularExpression) {
            let substr = String(text[match])
            if let numMatch = substr.range(of: #"\d{1,3}"#, options: .regularExpression),
               let val = Double(String(text[numMatch])) {
                return min(130, max(0, val))
            }
        }
        if let match = text.range(of: #"\d{1,3}%"#, options: .regularExpression) {
            let substr = String(text[match])
            let num = substr.replacingOccurrences(of: "%", with: "")
            if let val = Double(num) {
                return min(130, max(0, val))
            }
        }
        return nil
    }

    private static func parseVoltageLine(from text: String, prefix: String) -> (Double?, Double?) {
        let pattern = "\(prefix):\\s*([0-9.]+)\\s*([0-9]+)\\s*Amps?"
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return (nil, nil) }
        let nsText = text as NSString
        guard let m = regex.firstMatch(in: text, range: NSRange(location: 0, length: nsText.length)),
              m.numberOfRanges >= 2 else { return (nil, nil) }
        let vRange = m.range(at: 1)
        guard vRange.location != NSNotFound,
              let voltage = Double(nsText.substring(with: vRange)) else { return (nil, nil) }
        var amps: Double? = nil
        if m.numberOfRanges >= 3 {
            let aRange = m.range(at: 2)
            if aRange.location != NSNotFound {
                amps = Double(nsText.substring(with: aRange))
            }
        }
        return (voltage, amps)
    }

    private static func parseRint(from text: String) -> Double? {
        if let match = text.range(of: #"Rint:\s*([\d.]+)\s*Ohms?"#, options: .regularExpression) {
            let substr = String(text[match])
            let nums = substr.components(separatedBy: CharacterSet.decimalDigits.union(CharacterSet(charactersIn: ".")).inverted)
                .filter { !$0.isEmpty }
                .compactMap { Double($0) }
            if let first = nums.first {
                return first
            }
        }
        if let match = text.range(of: #"[\d.]++\s*Ohms?"#, options: .regularExpression) {
            let substr = String(text[match])
            let numStr = substr.components(separatedBy: CharacterSet.decimalDigits.union(CharacterSet(charactersIn: ".")).inverted)
                .joined()
            if let val = Double(numStr) {
                return val
            }
        }
        return nil
    }
}
